collect_app/src/main/java/org/odk/collect/android/formentry/BackgroundAudioViewModel.java

```diff
@@ -0,0 +1,154 @@
+
+    private final MutableLiveData<Boolean> isPermissionRequired = new MutableLiveData<>(false);
+
+    // These fields handle storing record action details while we're granting permissions
+    private final HashSet<TreeReference> tempTreeReferences = new HashSet<>();
+    private String tempQuality;
+
+    public BackgroundAudioViewModel(AudioRecorder audioRecorder, PreferencesProvider preferencesProvider, RecordAudioActionRegistry recordAudioActionRegistry, PermissionsChecker permissionsChecker) {
+        this.audioRecorder = audioRecorder;
+        this.preferencesProvider = preferencesProvider;
+        this.recordAudioActionRegistry = recordAudioActionRegistry;
+        this.permissionsChecker = permissionsChecker;
+
+        this.recordAudioActionRegistry.register((treeReference, quality) -> {
+            new Handler(Looper.getMainLooper()).post(() -> handleRecordAction(treeReference, quality));
+        });
+    }
+
+    @Override
+    protected void onCleared() {
+        recordAudioActionRegistry.unregister();
+    }
+
+    public LiveData<Boolean> isPermissionRequired() {
+        return isPermissionRequired;
+    }
+
+    public boolean isBackgroundRecordingEnabled() {
+        return preferencesProvider.getGeneralSharedPreferences().getBoolean(KEY_BACKGROUND_RECORDING, true);
+    }
+
+    public void setBackgroundRecordingEnabled(boolean enabled) {
+        if (!enabled) {
+            audioRecorder.cleanUp();
+        }
+
+        preferencesProvider.getGeneralSharedPreferences().edit().putBoolean(KEY_BACKGROUND_RECORDING, enabled).apply();
+    }
+
+    public boolean isBackgroundRecording() {
+        return audioRecorder.isRecording() && audioRecorder.getCurrentSession().getValue().getId() instanceof Set;
+    }
+
+    public void grantAudioPermission() {
+        isPermissionRequired.setValue(false);
+        startBackgroundRecording(tempQuality, new HashSet<>(tempTreeReferences));
+        
+        tempTreeReferences.clear();
+        tempQuality = null;
+    }
+
+    private void handleRecordAction(TreeReference treeReference, String quality) {
+        if (isBackgroundRecordingEnabled()) {
+            if (permissionsChecker.isPermissionGranted(Manifest.permission.RECORD_AUDIO)) {
+                if (isBackgroundRecording()) {
+                    RecordingSession session = audioRecorder.getCurrentSession().getValue();
+                    HashSet<TreeReference> treeReferences = (HashSet<TreeReference>) session.getId();
+                    treeReferences.add(treeReference);
+                } else {
+                    HashSet<TreeReference> treeReferences = new HashSet<>();
+                    treeReferences.add(treeReference);
+
+                    startBackgroundRecording(quality, treeReferences);
+                }
+            } else {
+                isPermissionRequired.setValue(true);
+                tempTreeReferences.add(treeReference);
+                this.tempQuality = quality;
```

Comment:
SUPER edge case but I think this should only happen if `tempQuality` is `null` so we only capture the first quality.

Reply:
Yeah good point. I'll add this in.